source/extensions/filters/http/ext_proc/ext_proc.cc

```diff
@@ -1237,26 +1242,50 @@ void Filter::onReceiveMessage(std::unique_ptr<ProcessingResponse>&& r) {
   case ProcessingResponse::ResponseCase::kRequestHeaders:
     setDecoderDynamicMetadata(*response);
     processing_status = decoding_state_.handleHeadersResponse(response->request_headers());
+    if (on_processing_response_ != nullptr) {
+      on_processing_response_->afterProcessingRequestHeaders(*response, processing_status,
```

Comment:
Can afterProcessingRequestHeaders() and afterProcessingResponseHeaders() be put under handleHeadersResponse()?   That way, there is not too much logic directly under this switch... case .... state machine code.

Same for the handleBodyResponse() and handleTrailerResponse().

Reply:
I agree that having new calls within the switch case doesn't look ideal, but I feel it's better than the alternative. The afterProcessing<Request/Response>Headers methods require the `processing_status` which is not available in one place within handleHeadersResponse. See the multiple return statements 
[here](https://github.com/envoyproxy/envoy/blob/main/source/extensions/filters/http/ext_proc/processor_state.cc#L146)
Moving the calls within handleHeadersResponse would make the code very complex and brittle.